src/lib/Libifl/pbsD_deljoblist.c

```diff
@@ -46,160 +46,471 @@
+ * @brief	Convenience function to send deljoblist request and receive reply
  *
- * @par Functionality:
- *              This request is used for delete operation.
- *              
+ * @param[in] fd - communication handle
+ * @param[in] function - req type
+ * @param[in] jobids - the job id list
+ * @param[in] numjids - the count of job ids
+ * @param[in] extend - extend string for req
  *
- * @param[in] c - socket descriptor
- * @param[in] command - command type
- * @param[in] objtype - object type
- * @param[in] objname - object name
- * @param[in] aoplp - pointer to attropl structure(list)
- * @param[in] prot - PROT_TCP or PROT_TPP
- * @param[in] msgid - message id
+ * @return struct batch_deljob_status *
+ * @retval list of error objects from server (or NULL)
+ */
+static struct batch_deljob_status *
+deljob_to_server(int fd, int function, char **jobids, int numids, char *extend)
+{
+	if (send_deljob(fd, function, jobids, numids, extend))
+		return NULL;
+
+	return recv_deljob(fd);
+}
+
+/**
+ * @brief	Broadcast deljob request to other servers in a multi-server set up
+ * 			This handles the cases where a job was moved to another server
  *
- * @return      int
- * @retval      DIS_SUCCESS(0)  success
- * @retval      error code      error
+ * @param[in] c - connection handle (virtual)
+ * @param[in] orig_fd - connection handle of the server which originally owned the job
+ * @param[in] function - req type
+ * @param[in] jobids - the job id list
+ * @param[in] numjids - the count of job ids
+ * @param[in] extend - extend string for req
  *
+ * @return int
+ * @retval 0 for Success
+ * @retval 1 for Error
  */
-int
-PBSD_deljoblist_put(int c, int function, char **jobids, int numofjobs, char *extend, int prot, char **msgid)
+static int
+broadcast_deljob(int c, int orig_fd, int function, char **jobids, int numids, char *extend)
 {
-	int rc;
+	int *skip_list;
+	svr_conn_t *conns = get_conn_svr_instances(c);
+	int job_found = 0;
+	int i;
+
+	if (conns == NULL) {
+		pbs_errno = PBSE_NOCONNECTS;
+		return 1;
+	}
 
-	if (prot == PROT_TCP) {
-		DIS_tcp_funcs();
-	} else {
-		if ((rc = is_compose_cmd(c, IS_CMD, msgid)) != DIS_SUCCESS)
-			return rc;
+	skip_list = calloc(get_num_servers(), sizeof(int));
+	if (skip_list == NULL) {
+		pbs_errno = PBSE_SYSTEM;
+		return 1;
 	}
 
-	if ((rc = encode_DIS_ReqHdr(c, function, pbs_current_user)) ||
-		(rc = encode_DIS_JobsList(c, jobids, numofjobs)) ||
-		(rc = encode_DIS_ReqExtend(c, extend))) {
-		if (prot == PROT_TCP) {
-			if (set_conn_errtxt(c, dis_emsg[rc]) != 0)
-				return (pbs_errno = PBSE_SYSTEM);
+	/* Broadcast it to all servers */
+	for (i = 0; i < get_num_servers(); i++) {
+		if (conns[i].sd > 0 && conns[i].sd != orig_fd) {
+			if (send_deljob(conns[i].sd, function, jobids, numids, extend))
+				skip_list[i] = 1;
+		} else
+			skip_list[i] = 1;
+	}
+	for (i = 0; i < get_num_servers(); i++) {
+		if (!skip_list[i]) {
+			struct batch_deljob_status *ret = NULL;
+
+			ret = recv_deljob(conns[i].sd);
```

Comment:
How does this work? You are sending a list of job ids to all other servers, what do we expect in recv_deljob? does each server reply per job or as a whole for the list of jobs?

Reply:
Each server replies only once, for the whole list of jobs. So, for 5 servers, we'll send 5 requests out and get 5 responses back